### DISCUSSION TITLE
.hex file size and Block Memory Generator question

### DISCUSSION THREAD

**Discussion by ZGCiou** [bol-edu/HLS-SOC-Discussions#183] [General]
我們在合併四種workload之後compile出的.hex file大小為14KB
超出FPGA上block memory generator的8KB
照我的理解這樣應該沒有辦法把完整的firmware放上FPGA運行
在FPGA上做驗證時也出現預期之外的結果
<img width="749" alt="截圖 2023-12-12 下午5 20 54" src="https://github.com/bol-edu/HLS-SOC-Discussions/assets/127314907/9ed8311a-27ea-40ba-9cbb-e805ac247536">
<img width="1148" alt="截圖 2023-12-12 下午4 20 18" src="https://github.com/bol-edu/HLS-SOC-Discussions/assets/127314907/8734b52e-461a-436b-985f-300bdba8933d">
應該出現AB61卻只停在AB51
我們猜測有可能和.hex超出8K有關

之後我有嘗試修改block memory generator的大小
<img width="1051" alt="截圖 2023-12-12 下午4 02 46" src="https://github.com/bol-edu/HLS-SOC-Discussions/assets/127314907/b78444d7-90d3-4267-9f5b-5b09fbbfdbca">
但是在vivado內size的欄位是固定無法修改
就算下指令修改 開始run synthesis或是validation design後又會變回8K

總結一下我的問題
1. .hex file 是否一定要小於8K
2. Block Memory Generator 的大小有沒有辦法調整, 如果不行是什麼原因造成的

不知道有沒有人遇到類似的問題或是能幫我解答疑問
感謝

**Comment by ZGCiou** [comment]
*update question 1

試了給compiler augment -Os 或 O1 可以壓在8KB以內，成功跑出firmware給的flag

**Comment by boledulab** [comment]
@ZGCiou Great!  This is a smart method.  In Lab 4.2, some teams use compiler optimization to push the loop code into CPU cache (64B instruction cache) and get pretty amazing results.

**Comment by Lin-Xiang88** [comment]
Block Memory Generator在Basic的地方把mode調成Stand Alone就可以調整大小了，不過調mode對於電路是否有影響要測試一下
![image](https://github.com/bol-edu/HLS-SOC-Discussions/assets/43726100/1542f5b4-a171-4144-8629-b52c1351f59f)

**Reply by ZGCiou** [reply]
好像少了reset pin，我再試試看有沒有影響
感謝你

**Reply by ZGCiou** [reply]
後來發現勾選32 bit address的話port和原本的不會有差異
也順利在FPGA上成功run 8KB以上的code了
感謝你
